apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerDataReader.cs

```diff
@@ -0,0 +1,349 @@
+        public Value GetJsonValue(int i)
+        {
+            return (Value) _innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(Value));
+        }
+
+        /// <inheritdoc />
+        public override string GetName(int i)
+        {
+            return _resultset.GetMetadataAsync(CancellationToken.None).Result.RowType.Fields[i].Name;
+        }
+
+        /// <inheritdoc />
+        public override int GetOrdinal(string name)
+        {
+            GaxPreconditions.CheckNotNullOrEmpty(name, nameof(name));
+            var fields = _resultset.GetMetadataAsync(CancellationToken.None).Result.RowType.Fields;
+            for (var i = 0; i < fields.Count; i++)
+                if (Compare(name, fields[i].Name, StringComparison.Ordinal) == 0)
+                    return i;
+            return -1;
+        }
+
+        /// <inheritdoc />
+        public override string GetString(int i)
+        {
+            return (string) _innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(string));
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="i"></param>
+        /// <returns></returns>
+        public Timestamp GetTimestamp(int i)
+        {
+            return (Timestamp) _innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(Timestamp));
+        }
+
+        /// <inheritdoc />
+        public override object GetValue(int i)
+        {
+            return this[i];
+        }
+
+        /// <inheritdoc />
+        public override int GetValues(object[] values)
+        {
+            var count = Math.Min(FieldCount, values.Length);
+            for (var i = 0; i < count; i++)
+                values[i] = this[i];
+            return count;
+        }
+
+
+        /// <inheritdoc />
+        public override bool IsDBNull(int i)
+        {
+            return GetJsonValue(i).KindCase == Value.KindOneofCase.NullValue;
+        }
+
+        /// <inheritdoc />
+        public override bool NextResult()
+        {
+            throw new NotSupportedException("Spanner does not support multiple SQL queries in a single command.");
+        }
+
+        /// <inheritdoc />
+        public override bool Read()
+        {
+            return ReadAsync(CancellationToken.None).Result;
+        }
+
+        /// <inheritdoc />
+        public override Task<bool> ReadAsync(CancellationToken cancellationToken)
+        {
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
+            {
+                await GetMetadataAsync(cancellationToken).ConfigureAwait(false);
+                _innerList.Clear();
+                //read # values == # fields.
+                var first = await _resultset.Next(cancellationToken).ConfigureAwait(false);
+                if (first == null)
+                    return false;
+                _innerList.Add(first);
+                //we expect to get full rows...
+                for (var i = 0; i < _metadata.RowType.Fields.Count - 1; i++)
```

Comment:
Possibly make that start at i = 1 and remove the -1? Keeps it a bit more consistent with "It's like the loop you'd expect, but we're starting at 1 because we've done the first iteration beforehand".